instrumentation/jdbc/library/src/main/java/io/opentelemetry/instrumentation/jdbc/OpenTelemetryDriver.java

```diff
@@ -0,0 +1,223 @@
+import io.opentelemetry.instrumentation.jdbc.internal.OpenTelemetryConnection;
+import java.sql.Connection;
+import java.sql.Driver;
+import java.sql.DriverManager;
+import java.sql.DriverPropertyInfo;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.Properties;
+import java.util.logging.Logger;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public class OpenTelemetryDriver implements Driver {
+
+  private static final String INTERCEPTOR_MODE_URL_PREFIX = "jdbc:otel:";
+
+  private static final OpenTelemetryDriver INSTANCE = new OpenTelemetryDriver();
+
+  private static boolean registered = false;
+  private static boolean interceptorMode = false;
+
+  static {
+    try {
+      register();
+    } catch (SQLException e) {
+      throw new ExceptionInInitializerError(e);
+    }
+  }
+
+  /**
+   * Ensure {@code TracingDriver} be the first driver of {@link DriverManager} to make sure
+   * "interceptor mode" works. WARNING: Driver like Oracle JDBC may fail since it's destroyed
+   * forever after deregistration.
+   */
+  public static synchronized void ensureRegisteredAsTheFirstDriver() {
+    try {
+      Enumeration<Driver> enumeration = DriverManager.getDrivers();
+      List<Driver> drivers = new ArrayList<>();
+      for (int i = 0; enumeration.hasMoreElements(); ++i) {
+        Driver driver = enumeration.nextElement();
+        if (i == 0 && driver == INSTANCE) {
+          // the first driver is OTEL driver, skip all this verification
+          return;
+        }
+        if (driver instanceof OpenTelemetryDriver) {
+          drivers.add(driver);
+        }
+        DriverManager.deregisterDriver(driver);
+      }
+
+      // register OTEL driver first
+      register();
+
+      // register other drivers
+      for (Driver driver : drivers) {
+        DriverManager.registerDriver(driver);
+      }
+    } catch (SQLException e) {
+      throw new ExceptionInInitializerError(e);
+    }
+  }
+
+  /**
+   * Turns "interceptor mode" on or off.
+   *
+   * @param interceptorMode The {@code interceptorMode} value.
+   */
+  public static void setInterceptorMode(final boolean interceptorMode) {
+    OpenTelemetryDriver.interceptorMode = interceptorMode;
+  }
+
+  /**
+   * Register the driver against {@link DriverManager}. This is done automatically when the class is
+   * loaded. Dropping the driver from DriverManager's list is possible using {@link #deregister()}
+   * method.
+   *
+   * @throws IllegalStateException if the driver is already registered
+   * @throws SQLException if registering the driver fails
+   */
+  public static void register() throws SQLException {
+    if (isRegistered()) {
+      throw new IllegalStateException(
+          "Driver is already registered. It can only be registered once.");
+    }
+    DriverManager.registerDriver(INSTANCE);
+    OpenTelemetryDriver.registered = true;
```

Comment:
`registered` should probably be an `AtomicBoolean` - for thread correctness

Reply:
Fixed. Thank you!